### DISCUSSION TITLE
Automate Quest List in OSM Wiki?

### DISCUSSION THREAD

**Discussion by kmpoppe** [streetcomplete/StreetComplete#3864] [Ideas]
Hey all,

I just played around with some basic PHP magic that allows me to make automated edits in the OSM Wiki.

As the [quest list](https://wiki.openstreetmap.org/wiki/StreetComplete/Quests) currently needs to be changed manually every time a quest gets changed or added, that list adds additional workload for the project.

If the meta-information that's listed in the Wiki was to be put into a .yml-File in the quest-source-folder, that could be the basis for each line in the Wiki Quest table - and putting this meta-information there could just be another task during quest development.

Additionally, creating and updating the language-pages (currently only EN, FR, JA exist) would get easier when some strings were coming from the resource-language-files.

Any thoughts on that?

Kai

**Comment by FloEdelmann** [comment]
See also #2523 (and the follow-up PRs linked at the bottom) for prior work.

**Reply by FloEdelmann** [reply]
There are some problems with the current manual approach, though; see e.g. #3887. More automation could avoid that.

**Comment by kmpoppe** [comment]
@FloEdelmann I might be too dense (or tired) ... how do I find the "lastest" quest-list.csv in the repo?

**Reply by FloEdelmann** [reply]
First, go to https://github.com/streetcomplete/StreetComplete/actions/workflows/generate-quest-list.yml:

<details><summary>Screenshot</summary>

![image](https://user-images.githubusercontent.com/202916/157955606-a4321da5-13b9-4ef5-8f87-38eb814ea009.png)

</details>

Then click on the latest run (https://github.com/streetcomplete/StreetComplete/actions/runs/1963886783):

<details><summary>Screenshot</summary>

![image](https://user-images.githubusercontent.com/202916/157955889-5df7a14c-f14e-4b7e-be38-bc530b9255af.png)

</details>

**Reply by kmpoppe** [reply]
Ah! Scrolling down to the artifacts list, would've helped before asking 🤦 Thanks!

**Comment by kmpoppe** [comment]
Just getting around to thinking about this a little more, @FloEdelmann, please take a look see at the quest-list.csv that generated with the last run https://github.com/streetcomplete/StreetComplete/actions/runs/2278485641. That (and also the one you pointed out in March) has three "empty" lines (only a bunch of commas) after the header-line.

Can you guess where they're coming from?

**Reply by FloEdelmann** [reply]
The CSV is generated here:

https://github.com/streetcomplete/StreetComplete/blob/18debe439abce6823a91b6f794c589a2a7be14a8/buildSrc/src/main/java/GenerateQuestListTask.kt#L185-L195

Since there are no `outdatedWikiQuests` and no `updatedRepoQuests`, there are three consecutive empty lines. I.e. the wiki is currently up-to-date with the repo.

**Reply by kmpoppe** [reply]
Could I get [this](https://github.com/kmpoppe/StreetComplete/commit/89b5fcd) change to the file (or something leading to a similar output) so that I can detect what QuestMap the entry comes from and I don't have to guess what to do with it?

**Reply by FloEdelmann** [reply]
Yes, but I think "Map Type" is a confusing title for it. That [`map`](https://kotlinlang.org/api/latest/jvm/stdlib/kotlin.collections/map.html) in the referenced code is executed on lists, so if anything, it should be "List Type". But even that is an implementation detail. If you're just looking at the generated CSV, you have no idea what "list" that is referring to. So either "Row Type" or just "Type".

Also, the column shouldn't be empty for rows in the first section (`outdatedWikiQuests`).

And I'd keep the empty lines, they make it easier to read when there are many changes at once, e.g. here: https://github.com/streetcomplete/StreetComplete/actions/runs/2240057997

**Reply by kmpoppe** [reply]
Changed to "type" and readded empty lines (ccebcea). For the outdated quests I added "???" at the end of val csvString in the wikiQuest class. Wouldn't that be enough to get a ??? in the last Column of outdatedWikiQuests?

**Reply by FloEdelmann** [reply]
Ah, you're right.

**Reply by kmpoppe** [reply]
> Ah, you're right.

Great ;) Will GH run the buildSrc as a test when I start a PR for that change?

I currently don't have the means to test if the result is as expected :-/

**Reply by FloEdelmann** [reply]
No, it will only run on commits on master. But you can enable GitHub Actions in your fork and merge it into your fork's master to test it.

**Reply by kmpoppe** [reply]
Neat! Thanks 👍
